File: client/src/main/java/com/lab/client/ClientApplication.java
package com.lab.client;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Leaderless replication client.
 *
 * Env vars:
 *   REPLICAS  – comma-separated list of replica base URLs
 *               e.g. http://replica1:8080,http://replica2:8080,http://replica3:8080
 */
public class ClientApplication {

    private static final Duration TIMEOUT = Duration.ofSeconds(5);

    public static void main(String[] args) throws Exception {
        String replicasEnv = System.getenv("REPLICAS");
        if (replicasEnv == null || replicasEnv.isBlank()) {
            System.err.println("REPLICAS environment variable is not set.");
            System.exit(1);
        }

        List<String> replicas = Arrays.stream(replicasEnv.split(","))
                .map(String::trim)
                .filter(s -> !s.isBlank())
                .toList();

        if (replicas.isEmpty()) {
            System.err.println("No replicas configured.");
            System.exit(1);
        }

        HttpClient http = HttpClient.newBuilder()
                .connectTimeout(TIMEOUT)
                .build();
        ObjectMapper mapper = new ObjectMapper();

        long ok = 0;
        long inconsistent = 0;
        long errors = 0;

        // Hide cursor and prepare for in-place output
        System.out.print("\033[?25l");
        Runtime.getRuntime().addShutdownHook(new Thread(() -> System.out.print("\033[?25h\n")));

        while (true) {
            // Step 1: pick a random replica, write a random value
            int writtenValue = ThreadLocalRandom.current().nextInt(Integer.MAX_VALUE);
            String writeReplica = randomPick(replicas);

            boolean writeOk = doWrite(http, mapper, writeReplica, writtenValue);
            if (!writeOk) {
                errors++;
                printStats(ok, inconsistent, errors);
                continue;
            }

            // Step 2: pick a random replica (possibly different), read back
            String readReplica = randomPick(replicas);
            Integer readValue = doRead(http, mapper, readReplica);
            if (readValue == null) {
                errors++;
                printStats(ok, inconsistent, errors);
                continue;
            }

            // Step 3: compare
            if (readValue == writtenValue) {
                ok++;
            } else {
                inconsistent++;
            }
            printStats(ok, inconsistent, errors);
        }
    }

    // -------------------------------------------------------------------------

    private static boolean doWrite(HttpClient http, ObjectMapper mapper,
                                   String replicaUrl, int value) {
        try {
            String body = mapper.writeValueAsString(new WriteBody(value));
            HttpRequest req = HttpRequest.newBuilder()
                    .uri(URI.create(replicaUrl + "/value"))
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .header("Content-Type", "application/json")
                    .timeout(TIMEOUT)
                    .build();
            HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
            return resp.statusCode() == 200;
        } catch (Exception e) {
            return false;
        }
    }

    private static Integer doRead(HttpClient http, ObjectMapper mapper, String replicaUrl) {
        try {
            HttpRequest req = HttpRequest.newBuilder()
                    .uri(URI.create(replicaUrl + "/value"))
                    .GET()
                    .timeout(TIMEOUT)
                    .build();
            HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
            if (resp.statusCode() == 200) {
                JsonNode node = mapper.readTree(resp.body());
                return node.get("value").asInt();
            }
        } catch (Exception ignored) {
        }
        return null;
    }

    private static String randomPick(List<String> list) {
        return list.get(ThreadLocalRandom.current().nextInt(list.size()));
    }

    private static void printStats(long ok, long inconsistent, long errors) {
        // \r moves to start of line to overwrite previous output
        System.out.printf("\rok: %d | inconsistent: %d | errors: %d   ", ok, inconsistent, errors);
        System.out.flush();
    }

    // Simple POJO for JSON serialisation
    static class WriteBody {
        public int value;
        WriteBody(int v) { this.value = v; }
    }
}

File: client/src/main/kotlin/com/lab/client/ClientApplication.kt
package com.lab.client

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit
import kotlin.random.Random

data class ValueVersion(val value: Int = 0, val version: Int = 0)

/**
 * Leaderless replication client.
 *
 * Env var:
 *   REPLICAS — comma-separated base URLs, e.g. http://replica1:8080,http://replica2:8080
 */
fun main() {
    val replicasEnv = System.getenv("REPLICAS")
        ?: error("REPLICAS environment variable is not set")

    val replicas = replicasEnv.split(",").map { it.trim() }.filter { it.isNotBlank() }
    check(replicas.isNotEmpty()) { "No replicas configured" }

    val http = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(5, TimeUnit.SECONDS)
        .writeTimeout(5, TimeUnit.SECONDS)
        .build()

    val mapper = jacksonObjectMapper()
    val json = "application/json".toMediaType()

    var ok = 0L
    var inconsistent = 0L
    var errors = 0L

    // Hide cursor for clean in-place output
    print("\u001B[?25l")
    Runtime.getRuntime().addShutdownHook(Thread { print("\u001B[?25h\n") })

    while (true) {
        // 1. Write to a random replica
        val writtenValue = Random.nextInt(Int.MAX_VALUE)
        val writeReplica = replicas.random()

        val writeBody = mapper.writeValueAsString(mapOf("value" to writtenValue))
            .toRequestBody(json)
        val writeReq = Request.Builder()
            .url("$writeReplica/value")
            .post(writeBody)
            .build()

        val writeOk = try {
            http.newCall(writeReq).execute().use { it.code == 200 }
        } catch (_: Exception) { false }

        if (!writeOk) {
            errors++
            printStats(ok, inconsistent, errors)
            continue
        }

        // 2. Read from a (possibly different) random replica
        val readReplica = replicas.random()
        val readReq = Request.Builder().url("$readReplica/value").get().build()

        val readValue = try {
            http.newCall(readReq).execute().use { resp ->
                if (resp.code == 200) {
                    mapper.readValue<ValueVersion>(resp.body!!.string()).value
                } else null
            }
        } catch (_: Exception) { null }

        if (readValue == null) {
            errors++
            printStats(ok, inconsistent, errors)
            continue
        }

        // 3. Compare
        if (readValue == writtenValue) ok++ else inconsistent++
        printStats(ok, inconsistent, errors)
    }
}

private fun printStats(ok: Long, inconsistent: Long, errors: Long) {
    print("\rok: $ok | inconsistent: $inconsistent | errors: $errors   ")
    System.out.flush()
}
